### PR TITLE
Port dilated_max_pool2d() to ATen

### PR DESCRIPTION
**Pull request by skrah**

### CODE REVIEW

**Comment by skrah** [comment]
Just as an explanation for the workaround, IntegrationTest.MNIST again passes an `IntArrayRef` of size `1`

```
May 20 03:01:34 C++ exception with description "max_pool2d_with_indices: internal error: all IntArrayRef sizes must be 2 (max_pool2d_with_indices_out_cuda_template at /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/DilatedMaxPool2d.cu:155)
```

**Comment by skrah** [comment]
@ezyang I think this is ready for review.

 I've deduplicated the shape checks in the original code, which in turn led to renaming the `Row/Col` convention in the `CUDA` code to `Height/Width` for compatibility with the `C++` code.

Hopefully it did not get out of hand and it's still reviewable. :)

File: aten/src/ATen/native/DilatedMaxPool.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    nInputPlane, inputHeight, inputWidth, outputHeight, outputWidth);
+
+  const int64_t ndim = input.ndimension();
+  const int64_t nOutputPlane = nInputPlane;
+
+  check_dim_size(gradOutput, ndim, ndim-3, nOutputPlane);
+  check_dim_size(gradOutput, ndim, ndim-2, outputHeight);
+  check_dim_size(gradOutput, ndim, ndim-1, outputWidth);
+
+  if (cuda) {
```
Comment (ezyang) [inline aten/src/ATen/native/DilatedMaxPool.h]: I... would believe you if you told me this is what the code did before, but this special case is kind of shocking XD

**Comment by ezyang** [comment]
> I've deduplicated the shape checks in the original code, which in turn led to renaming the Row/Col convention in the CUDA code to Height/Width for compatibility with the C++ code.

That is much appreciated, thank you!

File: aten/src/ATen/native/DilatedMaxPool.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+              "dilationH: ", dilationH, " dilationW: ", dilationW);
+
+  TORCH_CHECK(input.numel() > 0 && (ndim == 3 || ndim == 4),
+              "non-empty 3D or 4D input tensor expected but got ndim: ", ndim);
+  TORCH_CHECK(kW/2 >= padW && kH/2 >= padH,
+              "pad should be smaller than half of kernel size, but got ",
+              "padW = ", padW, ", padH = ", padH, ", kW = ", kW, ", kH = ", kH);
+
+  if (outputWidth < 1 || outputHeight < 1) {
+    AT_ERROR("Given input size: (",
```
Comment (ezyang) [inline aten/src/ATen/native/DilatedMaxPool.h]: nit: `TORCH_CHECK(outputWidth >= 1 && outputHeight >= 1, msg)`

File: aten/src/ATen/native/DilatedMaxPool2d.cpp [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+          Tensor& output,
+          Tensor& indices,
+          const Tensor& input_,
+          IntArrayRef kernel_size,
+          IntArrayRef stride,
+          IntArrayRef padding,
+          IntArrayRef dilation,
+          bool ceil_mode)
+{
+  // XXX JIT: Pooling.cpp allows stride.empty().
```
Comment (ezyang) [inline aten/src/ATen/native/DilatedMaxPool2d.cpp]: This sounds like another misannotated function.

File: aten/src/ATen/native/DilatedMaxPool.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+#pragma once
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename dest_t, typename src_t>
+static inline dest_t
+safe_downcast(src_t v)
```
Comment (ezyang) [inline aten/src/ATen/native/DilatedMaxPool.h]: Nice!

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/DilatedMaxPool.h (hunk 1 above)
Comment (skrah) [reply aten/src/ATen/native/DilatedMaxPool.h]: It *is* shocking, but it's indeed what the old code did. :)  The CUDA code here always uses 4 dimensions in the kernel for `output`, then resizes back after the kernel:

```
 if(input->dim() == 3)
    THCTensor_(resize3d)(state, output, nInputPlane, nOutputRows, nOutputCols);
```

The C code uses 3 or 4 dimensions for the kernel.


Perhaps the shape check should be *after* all resizes and just before the kernel, but the old code checked right at the top of the function.

File: aten/src/ATen/native/DilatedMaxPool2d.cpp (hunk 3 above)
Comment (skrah) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: The annotation `Optional[BroadcastingList2[int]]` looks correct, but there's some explicit handling of `stride` here:

https://github.com/pytorch/pytorch/blob/70ecddfd76e19e856f2a5aa61c2123eee611af7b/torch/nn/functional.py#L482

I wonder if that could just be:

```
if stride is None:
    stride = kernel_size
```

The drawback is that one could crash `cpp` code by modifying Python code, but that is also the case for the annotation.

Comment (skrah) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: The other special case is here:

https://github.com/pytorch/pytorch/blob/70ecddfd76e19e856f2a5aa61c2123eee611af7b/test/cpp/api/integration.cpp#L303

It ends up calling `max_pool2d_with_indices` with all default parameters.

I didn't look very hard, but I couldn't find the overload of `torch::max_pool2d(conv1->forward(x), {2, 2})` that works with just two parameters in the `cpp` files.

File: aten/src/ATen/native/DilatedMaxPool.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/DilatedMaxPool.h]: Yeah, I think that's the correct fix. No need to block this PR on fixing it, would be a nice readability improvement though.

File: aten/src/ATen/native/DilatedMaxPool2d.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: > The drawback is that one could crash cpp code by modifying Python code, but that is also the case for the annotation.

I think, for now, we should consider these annotations as part of the TCB. I do wonder a little if they can't be checked for consistency with the main annotations, but... well... someone would have to figure that out :>

**Comment by ezyang** [comment]
You forgot to delete `aten/src/THCUNN/SpatialDilatedMaxPooling.cu` and it's thus failing internal contbuilds. You didn't allow maintainers to edit your branch so I can't push the fix. Can you please fix this? :)

**Review by ezyang — CHANGES_REQUESTED** [review]
Delete file

File: aten/src/ATen/native/DilatedMaxPool2d.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: > I wonder if that could just be:

Then we'd have to explain why it was permissible for all the other sites to also default things to empty list. Maybe they're all wrong.

@ailzhang, you helped us resolve the annotation last time in #20306. Do you know what's going on with these `Optional[BroadcastingList2]` things?

**Comment by skrah** [comment]
You didn't allow maintainers to edit your branch so I can't push the fix. Can you please fix this? :)

Done, I'll leave the "allow" button on next time, it's just a habit of mine from other projects :)

File: aten/src/ATen/native/DilatedMaxPool2d.cpp (hunk 3 above)
Comment (ailzhang) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: Hi @skrah @ezyang , yea this annotation looks fine, but I guess JIT should have instantiate it to int[2] before passing into aten. I'm looking into it, to double check the repro process:
1. checking out this PR
2. Run IntegrationTest.MNIST should trigger stride.empty() here right?
Thanks!

Comment (skrah) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: Hi @ailzhang, I think the `stride.empty()` case is also triggered by the regular Python tests if that is more convenient.

`IntegrationTest.MNIST` triggers all of `stride.empty() && padding.size()==1 && dilation.size()==1`.

Comment (skrah) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: And (for the sake of completeness) this PR already contains a workaround, so `assert()` will need to be added to reproduce it.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/DilatedMaxPool2d.cpp (hunk 3 above)
Comment (ailzhang) [reply aten/src/ATen/native/DilatedMaxPool2d.cpp]: @ezyang I have a commit to fix this(and other similar ones). Once this PR is merged into master, I will rebase and send it out. :D

**Pull request closed without merging**